compiler/rustc_query_impl/src/plumbing.rs

```diff
@@ -106,28 +106,32 @@ impl QueryContext for QueryCtxt<'_> {
         diagnostics: Option<&Lock<ThinVec<Diagnostic>>>,
         compute: impl FnOnce() -> R,
```

Comment:
Did you try to make `start_query` non-generic by passing a `&mut dyn FnMut()`?

Reply:
I don't think that will have as much of an effect - the return type still needs to be generic.

Reply:
and actually I'm not sure it's possible to do at all - using `FnMut` means we can't capture `key` in the caller :/ we need `&owned` or something like that.
```
error[E0507]: cannot move out of `key`, a captured variable in an `FnMut` closure
   --> compiler/rustc_query_system/src/query/plumbing.rs:470:63
    |
419 |     key: Q::Key,
    |     --- captured outer variable
...
458 |         qcx.start_query(job_id, Q::DEPTH_LIMIT, Some(&diagnostics), &mut || {
    |                                                                          -- captured by this `FnMut` closure
...
470 |             dep_graph.with_task(dep_node, *qcx.dep_context(), key, task, Q::HASH_RESULT)
    |                                                               ^^^ move occurs because `key` has type `<Q as QueryConfig<Qcx>>::Key`, which does not implement the `Copy` trait
```

Reply:
I'm wrong, sorry, `key` can be made Copy.